backends/edk2-compat/edk2-svc-generate.c

```diff
@@ -232,200 +206,148 @@ int performGenerateCommand(int argc,char* argv[])
-					rc = ARG_PARSE_FAIL;
-					goto out;
-				}
-				else {
-					i++;
-					args->signKeyCount++;
-					args->signKeys = realloc(args->signKeys, args->signKeyCount * sizeof(char*));
-					args->signKeys[args->signKeyCount - 1] = argv[i];
-				}
-			}
-            // set signed data, alternative to -k   
-            else if (!strcmp(argv[i], "-s")) {
-                //if already storing private keys, then don't allow for signed data
-                if (args->alreadySignedFlag == 0){
-                    prlog(PR_ERR, "ERROR: Cannot have both signed data files and private keys for signing");
-                    goto out;
-                }
-                args->alreadySignedFlag = 1;
-                if (i + 1 >= argc || argv[i + 1][0] == '-') {
-                    prlog(PR_ERR, "ERROR: Incorrect signed data flag, see usage...\n");
-                    rc = ARG_PARSE_FAIL;
-                    goto out;
-                }
-                else {
-                    i++;
-                    args->signKeyCount++;
-                    args->signKeys = realloc(args->signKeys, args->signKeyCount * sizeof(char*));
-                    args->signKeys[args->signKeyCount - 1] = argv[i];
-                }
+	}
+
+	switch (key) {
+		case 'k':
+			 //if already storing signed data, then don't allow for private keys
+            if (args->alreadySignedFlag == 1){
+                prlog(PR_ERR, "ERROR: Cannot have both signed data files and private keys for signing\n");
+                rc = ARG_PARSE_FAIL;
+                break;
             }
-			// set public key signer
-			else if(!strcmp(argv[i], "-c")) {	
-				if (i + 1 >= argc || argv[i + 1][0] == '-') {
-					prlog(PR_ERR, "ERROR: Incorrect value for public key flag, use 'c <cert>', see usage...\n");
-					rc = ARG_PARSE_FAIL;
-					goto out;
-				}
-				else {
-					i++;
-					args->signCertCount++;
-					args->signCerts = realloc(args->signCerts, args->signCertCount * sizeof(char*));
-					args->signCerts[args->signCertCount - 1] = argv[i];
-				}
-			}
-			// set input file
-			else if (!strcmp(argv[i], "-i")) {
-				if (i + 1 >= argc || argv[i + 1][0] == '-') {
-					prlog(PR_ERR, "ERROR: Incorrect flag '-i', see usage...\n");
-					rc = ARG_PARSE_FAIL;
-					goto out;
-				}
-				else {
-					i++;
-					args->inFile = argv[i];
-				}
-			}
-			// set output file 
-			else if (!strcmp(argv[i], "-o")) {
-				if (i + 1 >= argc || argv[i + 1][0] == '-') {
-					prlog(PR_ERR, "ERROR: Incorrect flag '-o', see usage...\n");
-					rc = ARG_PARSE_FAIL;
-					goto out;
-				}
-				else {
-					i++;
-					args->outFile = argv[i];
-				}
+            args->alreadySignedFlag = 0;
+			args->signKeyCount++;
+			args->signKeys = realloc(args->signKeys, args->signKeyCount * sizeof(char*));
+			args->signKeys[args->signKeyCount - 1] = arg;
+			break;
+		case 'c':
+			args->signCertCount++;
+			args->signCerts = realloc(args->signCerts, args->signCertCount * sizeof(char*));
```

Comment:
Sadly, realloc can fail and return NULL. Worse, cleanup after realloc failure is _intensely_ annoying, as it returns NULL but does *not* free the memory you passed in. (See `man realloc`.)

There's also the (**entirely** theoretical IMO) issue that the multiplication here might overflow. `reallocarray` would handle that for you but that's another gnu extension.

not sure how best to get around this, I'd say probably a wrapper function but I'm not fussy.

(btw, it occured to me that maybe if you have gcc10 you can get the compiler to find stuff like this for you: try `-fanalyzer` per https://developers.redhat.com/blog/2020/03/26/static-analysis-in-gcc-10/ Currently it finds heaps of bugs, none of which I have triaged, but it looks like at least some are in your arg parser code, so maybe worth a look at some point, but I'd advise against fixing unrelated issues in this series, save it for a followup.)

Reply:
Thanks for pointing this out. Interesting realloc can return NULL without freeing. In response to this, I made a new commit that uses a new function to to realloc the array. I will add to this PR (sorry!) for sake of less PR's and getting things moving.

Reply:
As for using `-fanalyzer` , I have added updating my compiler and using the flag to the secvarctl trello board.

Reply:
Cool, I'll check that out. yes, realloc has an 'interesting' calling convention. I'll ping you internally about the trello board, I'd be interested in seeing what your grand plans are :)